agent/testflinger_agent/agent.py

```diff
@@ -179,168 +116,68 @@ def mark_device_offline(self):
-                        phase, rundir
-                    )
-                    self.client.post_influx(phase, exit_code)
-                    event_emitter.emit_event(exit_event, exit_reason)
-                    detail = ""
-                    if exit_code:
-                        # exit code 46 is our indication that recovery failed!
-                        # In this case, we need to mark the device offline
-                        if exit_code == 46:
-                            self.mark_device_offline()
-                            exit_event = TestEvent.RECOVERY_FAIL
-                        else:
-                            exit_event = TestEvent(phase + "_fail")
-                        detail = parse_error_logs(error_log_path, phase)
-                    else:
-                        exit_event = TestEvent(phase + "_success")
-                    event_emitter.emit_event(exit_event, detail)
-                    if phase == TestPhase.PROVISION:
-                        self.client.post_provision_log(
-                            job.job_id, exit_code, exit_event
-                        )
-                    if exit_code and phase != TestPhase.TEST:
-                        logger.debug("Phase %s failed, aborting job" % phase)
-                        job_end_reason = exit_event
-                        break
-            except Exception as e:
-                logger.exception(e)
+                    # Run the phase or skip it
+                    if job.go(phase):
+                        self.set_agent_state(phase)
+                        job.run(phase)
+                        if job.check_end():
+                            break
+            except Exception as error:
+                logger.exception(f"{type(error).__name__}: {error}")
             finally:
+                phase = TestPhase.CLEANUP
                 # Always run the cleanup, even if the job was cancelled
-                event_emitter.emit_event(TestEvent.CLEANUP_START)
-                job.run_test_phase(TestPhase.CLEANUP, rundir)
-                event_emitter.emit_event(TestEvent.CLEANUP_SUCCESS)
-                event_emitter.emit_event(TestEvent.JOB_END, job_end_reason)
-                # clear job id
-                self.client.post_agent_data({"job_id": ""})
+                if job.go(phase):
+                    self.set_agent_state(phase)
+                    job.run(phase)
+
+            # let the server know the agent is available (clear job id)
+            job.end()
+            self.client.post_agent_data({"job_id": ""})
+
+            provision_result = job.phases[TestPhase.PROVISION].result
+            if (
+                provision_result is not None
+                and provision_result.event == TestEvent.RECOVERY_FAIL
+            ):
+                self.mark_device_offline()
 
             try:
-                self.client.transmit_job_outcome(rundir)
+                self.client.transmit_job_outcome(str(job.params.rundir))
```

Comment:
So I forgot that I had the edge revision of serve from Varun's branch deployed when I was testing this today, and predictably hit a 422 error when it tried to transmit the outcome of a job containing attachments. But then I realized the job was never marked complete. This is because that information is conveyed through transmit_job_outcome. The result is that the CLI is just left in a "stuck" state waiting for the job to complete.

Of course, this isn't really from your patch, but it gave me an "aha" moment for this pathological case that could absolutely occur. I wonder if instead of just failing here, if we could go ahead an try to at least do a client.post_job_state() with the state that we intend to transmit (cancelled or complete) if we fall into the exception block here? That would help with things like this in the future.